docs/contributing/999-rfcs/2024-02-15-resource-pinning.md

```diff
@@ -0,0 +1,239 @@
+
+Running `wing compile -t tf-aws` also succeeds with the following output and a new file called main.w.pin.json appears:
+
+```shell
+# (output)
+[+pin] /root/CoolBucket
+```
+
+```json
+// main.w.pin.json
+{
+  "version": "1",
+  "pinned": {
+    "tf-aws": {
+      "/root/CoolBucket": {
+        "type": "aws_s3_bucket",
+        "uniqueId": "CoolBucket",
+        "path": "/root/CoolBucket",
+      }
+    }
+  }
+}
+```
+
+Sally deploys this to AWS and everything is great.
+
+Some time later, Sally decided to move the bucket into a new abstraction.
+
+```wing
+// main.w
+bring cloud;
+
+class MyBucket {
+  bucket: cloud.Bucket;
+
+  new() {
+    this.bucket = new cloud.Bucket();
+  }
+}
+
+let myBucket = new MyBucket();
+
+```
+
+`wing compile` and `wing test` still work as expected.
+
+`wing compile -t tf-aws` however fails with an error message
+
+```shell
+# (output)
+[-pin] /root/CoolBucket
+[+pin] /root/MyBucket/cloud.Bucket
+
+Error: The following pin has been moved or deleted:
+  - /root/CoolBucket
+
+If moved, update "path" property in the pinfile to the new path.
+If deleted, remove the entry from the pinfile.
+```
+
+For now, Sally wants to avoid the destructive move of the bucket so she updates the `path` in the pinfile.
+
+```json
+// main.w.pin.json
+{
+  "version": "1",
+  "pinned": {
+    "tf-aws": {
+      "/root/CoolBucket": {
+        "type": "aws_s3_bucket",
+        "uniqueId": "CoolBucket",
+        "path": "/root/MyBucket/cloud.Bucket"
+      }
+    }
+  }
+}
+```
+
+Running `wing compile -t tf-aws` now succeeds and nothing is deleted or moved in the resulting compiled output.
+
+Later, Sally decides the bucket has no need to be so protected, so they remove the default pin on `cloud.Bucket`:
+
+```wing
+// main.w
+bring cloud;
+
+class MyBucket {
+  bucket: cloud.Bucket;
+
+  new() {
+    this.bucket = new cloud.Bucket();
+    this.bucket.pinned = false;
+  }
+}
+
+let myBucket = new MyBucket();
+```
+
+Running `wing compile -t tf-aws` automatically updates the pinfile (because the resource is otherwise unchanged):
+
+```shell
+# (output)
+[-pin] /root/CoolBucket
+```
+
+## Implementation
+
+### pinning
+
+The `pinned` property is used to mark a created resource as "pinned" in the resource tree. This is an inherent public property of all resources. This means that the user does not intend for the resource to be deleted or moved. Pinning is deep, meaning that pinning a parent inherently pins all of its children.
+
+This information will be available to the target platform/app which can choose what to do with this information. For example, the sim target can simply ignore this while the terraform target can manage a pinfile to inform the user when this changes between compilation and allow them to resolve any renames.
+
+### Annotated Pinfile
+
+```json
+{
+  "version": "1",
+  "pinned": {
+    "tf-aws": {
+      // The original logical path to the resource. These will only be leaves of the resource tree
```

Comment:
not sure I understand why we need the "original" logical path? what's the value?

Reply:
Doesn't make sense to me to keep the original path once the resource is moved. We can change the mapping to be physical -> logical instead of logical -> physical and then it might make more sense.

Reply:
Why not both?

> Even if we have tf attribute overrides, having an easy remapping is useful.

I think it's pretty common to want to logically remap, why not allow it alongside physical Instead of forcing physical remappings (there could be multiple for anything derived by the path)

Reply:
Logical remapping is a pretty different use case and I "smell" that there are quite a lot of implications to it that we need to think about. I recommend *not* including it in this feature at the moment.

Reply:
I have now removed any logical-path aspect of this feature